### PR TITLE
Refactor HSTSList to use HashMap

### LINKED ISSUE #14756: Use another backing store for the HSTS list.
Seems pretty inefficient to use a `Vec` there, when we could index by domain and then check subdomains.

This shouldn't be extremely hard to do, and there's a TODO comment there that mentions this.

Seems that the failures in #14716 weren't related to this, but seems worth to do.

### PR DESCRIPTION
**Pull request by nmvk**
Refactored HSTSList to use HashMap, where the key of HashMap is the base domain. Every time when we check if a host is secure, we find the base domain of the host and get a vector of HSTS entries associated with the base domain. 

While this will not give O(1) look up time, we would have a smaller list to iterate for every lookup. I have added one unit test to validate `HashMap` changes.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14756

- [X] There are tests for these changes OR

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14794)

### CODE REVIEW

File: components/net/hsts.rs [hunk 1]
```diff
@@ -78,40 +94,46 @@ impl HstsList {
...
-        // Could optimise by searching for exact matches first (via a map or
-        // something), then checking for subdomains.
-        self.entries.iter().any(|e| {
-            if e.include_subdomains {
-                e.matches_subdomain(host) || e.matches_domain(host)
-            } else {
-                e.matches_domain(host)
+        let base_domain = reg_host(host);
+
+        match self.entries_map.get(&base_domain) {
```
Comment (emilio) [inline components/net/hsts.rs]: nit: You can use:

```rust
self.entries_map.get(&base_domain).map_or(false, |entries| {
   // ...
})
```

To save a bit of indentation. Also applies below.

Comment (mrnayak) [inline components/net/hsts.rs]: Thanks!

File: components/net/hsts.rs [hunk 2]
```diff
@@ -132,3 +148,7 @@ pub fn secure_url(url: &Url) -> Url {
         url.clone()
     }
 }
+
+fn reg_host<'a>(url: &'a str) -> String {
```
Comment (emilio) [inline components/net/hsts.rs]: I don't think this is needed? Just use `reg_suffix`, and when needed do `to_owned` or similar, but you don't need a `String` all the time.

**Review by emilio — APPROVED** [review]
Looks good to me once those comments are addressed, thanks for doing this! :)

File: components/net/hsts.rs (hunk 2 above)
Comment (mrnayak) [inline components/net/hsts.rs]: Thanks, I had to add `let host = entry.host.clone();` to solve borrowing issue is this fine?

**Comment by emilio** [comment]
@bors-servo r+

Yeah, it's ok.

Thanks for doing this! :)

**Comment by mrnayak** [comment]
Thanks @emilio

**Pull request merged**